### PR TITLE
Fix more corner cases in stdlib migrator

### PR DESCRIPTION
**Pull request by h-vetinari**
Addresses #2322 and the discussion in https://github.com/conda-forge/htcondor-feedstock/pull/225

### CODE REVIEW

File: tests/test_yaml/stdlib_pagmo_after_meta.yaml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+outputs:
+  - name: pagmo
+    requirements:
+      host:
+        - eigen
+        - nlopt
+        - libboost-devel
+        - tbb-devel
+        - ipopt
```
Comment (h-vetinari) [inline tests/test_yaml/stdlib_pagmo_after_meta.yaml:33]: So this is a case where I'm not sure that the "no `script:`, no `build:`" rule creates the correct result, as I have often experienced that conda does _not_ propagate run-exports from the global build environment to the outputs. 

This particular case is even _more_ special because it does the "same name for feedstock and output" thing (which I consider highly fragile in practice), so it actually might work, but the question to me remains: are we 100% sure we will not create missing run-exports to `__glibc / __osx` if we _don't_ add `{{ stdlib("c") }}` in such a case?

**Review by h-vetinari — COMMENTED** [review]
It's interesting that the two cases brought up actually interact with each other, which makes me wonder if the rules we're using to detect "pure metapackages" is robust enough.

Would very much like your opinions here @mbargull @beckermr

**Comment by isuruf** [comment]
I haven't followed the code here, but isn't it enough to add `stdlib` only to places where `compiler` already exists?

**Comment by h-vetinari** [comment]
> I haven't followed the code here, but isn't it enough to add `stdlib` only to places where `compiler` already exists?

That's exactly the question I'm trying to ask in that review comment. From past experience[^1], having a multi-output recipe where only the global stage has a `build:` section, does not (or not always?) forward the run-exports generated by that global build section to the outputs that are assembled from the pieces being produced by the global stage.

In that case, to ensure that the outputs themselves have the right run-exports, they need to have the respective host (in this case build:) dependence themselves.

We did the same thing for the boost-migrator - if the global host-section contained `boost-cpp`, we added `libboost-devel` as a host-dep to the outputs, to ensure that the run-exports get set up correctly.

If we can leave that out - all the better, but I was less concerned about "overdoing" it, than about creating outputs with broken metadata.

[^1]: it's very possible that I haven't fully understood how this is supposed to work; I remember having had problems with this, though perhaps there had been other factors at play as well.

**Comment by isuruf** [comment]
> That's exactly the question I'm trying to ask in that review comment. From past experience[1](https://github.com/regro/cf-scripts/pull/2327#user-content-fn-1-3a83c9f78952c343b3231c81802e189e), having a multi-output recipe, where the only non-trivial build step happens during the "global" stage (i.e. there's only one build:/host: pair that is supposed to generate the right run-exports), does not (or not always?) forward these run-exports to the outputs that are assembled from the pieces being produced by the global stage.

That's not exactly relevant right? We should assume that the recipe is correct and if it is correct, we only need to add `stdlib` to sections where there is a `compiler`.

**Comment by h-vetinari** [comment]
I don't see how it _cannot_ be relevant. If you look at that pagmo example and it indeed turns out that run-exports aren't forwarded to the outputs, then none of the artefacts produced by that feedstock would pick up constraints on `__glibc / __osx`. That would mean that as soon as we increase the baseline stdlib version, we'd be creating packages with incorrect constraints, which seems to me like it would defeat the entire purpose of the stdlib-effort.

**Comment by h-vetinari** [comment]
I mean, if we're willing to tell such feedstocks: "you're doing it wrong, now you silently produce artefacts with incorrect metadata", then sure, the implementation becomes much simpler.

The question is where we draw the line in terms of saying that it's our responsibility to ensure that feedstocks that are currently "working", while actually ignoring compiler-exports, are fixed during this migration.

**Comment by beckermr** [comment]
Let's merge and keep going. The edge cases are numerous. Our feedstocks will find them.

**Comment by isuruf** [comment]
> I don't see how it cannot be relevant. If you look at that pagmo example and it indeed turns out that run-exports aren't forwarded to the outputs, then none of the artefacts produced by that feedstock would pick up constraints on __glibc / __osx. That would mean that as soon as we increase the baseline stdlib version, we'd be creating packages with incorrect constraints, which seems to me like it would defeat the entire purpose of the stdlib-effort.

pagmo recipe is wrong. It should have had the compiler in the build case. It is already wrong because the package doesn't have `libgcc-ng` and `libstdcxx-ng` in run due to the absence of `compiler`.

> you're doing it wrong, now you silently produce artefacts with incorrect metadata", 

I think we are willing to say, "your metadata is wrong already and there will be more missing metadata".

**Comment by h-vetinari** [comment]
> pagmo recipe is wrong. It should have had the compiler in the build case. It is already wrong because the package doesn't have `libgcc-ng` and `libstdcxx-ng` in run due to the absence of `compiler`.

That's fair enough, though it realistically worked without issue, because every non-trivial environment has `libgcc-ng` and `libstdcxx-ng`

> I think we are willing to say, "your metadata is wrong already and there will be more missing metadata".

I'm OK with that; it'll just make the stdlib migration more disruptive IMO (for the gain of cleaner recipes). In any case, I suggest to continue the discussion in #2328

**Comment by mbargull** [comment]
> I haven't followed the code here,

Same

> but isn't it enough to add `stdlib` only to places where `compiler` already exists?

I agree.
We shouldn't try to drive-by-fix too many things; there are too many (unintentionally broken or just *seemingly* broken) cases which deems an approach to address many things by some "simple-ish" modification to fail (and make undesired modifications in some cases).
IMHO, we should just look for `compiler` and possibly `sysroot_linux*`(mabye `__osx_`+`MACOSX_DEPLOYMENT_TARGET` in `run`/`run_constrained`?) but nothing more.
Meaning, adding a `requirements/build` section that hasn't been there before shouldn't be needed (unless you'd handle the `__osx` in `run`/`run_constrained` case).

> pagmo recipe is wrong. It should have had the compiler in the build case. It is already wrong because the package doesn't have `libgcc-ng` and `libstdcxx-ng` in run due to the absence of `compiler`.

The `pagmo` recipe is indeed not wrong.
It has a `pagmo` output without a `requirements/build` and as such `conda-build` will use the top-level `requirements/build` for that one (with the latter having the `compiler`s)
cf. https://github.com/conda/conda-build/blob/24.3.0/conda_build/metadata.py#L890-L894 and https://github.com/conda/conda-build/blob/24.3.0/conda_build/metadata.py#L1035-L1048 .
If you'd inspect `pagmo`'s dependencies, you should find the `run_export`ed libs.
